src/recover.rs

```diff
@@ -0,0 +1,294 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    collections::{HashSet, VecDeque},
+    sync::Arc,
+};
+
+use log::{debug, warn};
+use rayon::prelude::*;
+
+use crate::{
+    engine::MemTableAccessor, file_system::Readable, log_batch::LogItemBatch, memtable::MemTable,
+    reader::LogItemBatchFileReader, util::HashMap, FileId, GlobalStats, LogQueue, RecoveryMode,
+    Result,
+};
+
+struct RecoverContext {
+    file_id: FileId,
+    file_size: usize,
+    file_reader: Option<Box<dyn Readable>>,
+}
+
+impl std::fmt::Debug for RecoverContext {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("RecoverContext")
+            .field("file_id", &self.file_id)
+            .field("file_size", &self.file_size)
+            .finish()
+    }
+}
+
+#[derive(Default)]
+pub struct ParallelRecoverContext {
+    pub removed_memtables: HashSet<u64>,
+    pub compacted_memtables: HashMap<u64, u64>,
+    pub removed_keys: HashSet<(u64, Vec<u8>)>,
+}
+
+/// ParallelRecoverContext only masks previous items, not current or future items.
+impl ParallelRecoverContext {
+    pub fn merge(&mut self, right: &mut Self) {
+        self.removed_memtables
+            .extend(right.removed_memtables.drain());
+        self.compacted_memtables
+            .extend(right.compacted_memtables.drain());
+        self.removed_keys.extend(right.removed_keys.drain());
+    }
+}
+
+pub struct RecoverManager<F>
```

Comment:
Still not satisfied of this arrangement. How about abstracting the memtable recover logic into this:
```
    trait ReplayMachine : Default {
        fn replay(item_batch: LogItemBatch) -> Result<()>;
        fn merge(rhs: ReplayMachine) -> Result<()>;
    }
```
And let pipe log decides how to parallelize the recovery.

Reply:
I agree that the current design can still be improved, but let me  explain that why I choose it.

The basic rule is to avoid `file_pipe_log` and `memtable` rely on each other or rely on `engine`. But to make recovery parallel and under good performance, the recovery worker needs to:
1. Create new `MemTableAccessor` during parallel "reduce". Rayon will pick the appropriate points to insert an new empty `MemTableAccessor` to make "reduce" parallel.
2.  Create `ParallelRecoverContext` from `LogItemBatch`. Parallel recovery needs to get tombstones from item batch to mask previous memtables.

So, I think it's necessary to introduce `recover` mod and let mod `log_batch`, `memtable`, `file_pipe_log` and `engine` rely on it rather than on each other.

I don't think introduce `SequentialReplayMachine` will eliminate the dependency problem. But define it in `recover` mod and use it to separate functions for different usages is a good idea. But still, the information that recovery needs is not enough in `MemTableAccessor` (e.g. tombstones, currently in `ParallelRecoverContext`), and I don't want `MemTableAccessor` hold date that only for recovery. I'm not sure if GAT fits this condition.

Reply:
The `SequentialReplayMachine::default()` could be used to "reduce"? And the machine can holds both `MemTable` and `ParallelRecoverContext`.